File: overlay/scripts/collect-licenses.go
package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"io"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
)

var unsafeName = regexp.MustCompile(`[^A-Za-z0-9._+-]+`)

func safeName(value string) string {
	return strings.Trim(unsafeName.ReplaceAllString(value, "_"), "_")
}

func copyFile(source, destination string) error {
	in, err := os.Open(source)
	if err != nil {
		return err
	}
	defer in.Close()
	if err := os.MkdirAll(filepath.Dir(destination), 0755); err != nil {
		return err
	}
	out, err := os.Create(destination)
	if err != nil {
		return err
	}
	if _, err = io.Copy(out, in); err != nil {
		out.Close()
		return err
	}
	return out.Close()
}

func copyLicenseFiles(source, destination string) ([]string, error) {
	var files []string
	for _, pattern := range []string{"LICENSE*", "LICENCE*", "COPYING*", "NOTICE*"} {
		matches, err := filepath.Glob(filepath.Join(source, pattern))
		if err != nil {
			return nil, err
		}
		files = append(files, matches...)
	}
	sort.Strings(files)
	var copied []string
	for _, sourceFile := range files {
		info, err := os.Stat(sourceFile)
		if err != nil || !info.Mode().IsRegular() {
			continue
		}
		target := filepath.Join(destination, filepath.Base(sourceFile))
		if err := copyFile(sourceFile, target); err != nil {
			return nil, err
		}
		copied = append(copied, target)
	}
	return copied, nil
}

func main() {
	sourceFlag := flag.String("source", "", "Stratux source directory")
	outputFlag := flag.String("output", "", "License output directory")
	flag.Parse()
	if *sourceFlag == "" || *outputFlag == "" {
		fmt.Fprintln(os.Stderr, "source and output are required")
		os.Exit(2)
	}
	source, _ := filepath.Abs(*sourceFlag)
	output, _ := filepath.Abs(*outputFlag)
	if err := os.RemoveAll(output); err != nil {
		panic(err)
	}
	if err := os.MkdirAll(output, 0755); err != nil {
		panic(err)
	}

	index := []string{"Stratux NX bundled license files", ""}
	direct := map[string]string{
		"stratux-upstream":   filepath.Join(source, "LICENSE"),
		"stratux-nx":         filepath.Join(source, "debian", "stratux-nx-license"),
		"stratux-nx-notices": filepath.Join(source, "debian", "STRATUX-NX-NOTICES.md"),
	}
	directNames := []string{"stratux-upstream", "stratux-nx", "stratux-nx-notices"}
	for _, name := range directNames {
		path := direct[name]
		if _, err := os.Stat(path); err != nil {
			panic(fmt.Errorf("required license file is missing: %s", path))
		}
		target := filepath.Join(output, name+"-"+filepath.Base(path))
		if err := copyFile(path, target); err != nil {
			panic(err)
		}
		index = append(index, name+": "+filepath.Base(target))
	}

	for _, relative := range []string{"dump978", "dump1090", "rtl-ais", "ogn/ogn-tracker"} {
		destination := filepath.Join(output, safeName(relative))
		copied, err := copyLicenseFiles(filepath.Join(source, relative), destination)
		if err != nil {
			panic(err)
		}
		if len(copied) > 0 {
			index = append(index, relative+": "+filepath.Base(destination)+"/")
		}
	}

	command := exec.Command("go", "list", "-m", "-json", "all")
	command.Dir = source
	result, err := command.Output()
	if err != nil {
		panic(err)
	}
	decoder := json.NewDecoder(strings.NewReader(string(result)))
	moduleCount := 0
	for {
		var module struct {
			Path string
			Dir  string
			Main bool
		}
		if err := decoder.Decode(&module); err == io.EOF {
			break
		} else if err != nil {
			panic(err)
		}
		if module.Main || module.Path == "" || module.Dir == "" {
			continue
		}
		destination := filepath.Join(output, "go-modules", safeName(module.Path))
		copied, err := copyLicenseFiles(module.Dir, destination)
		if err != nil {
			panic(err)
		}
		if len(copied) > 0 {
			moduleCount++
			index = append(index, module.Path+": go-modules/"+filepath.Base(destination)+"/")
		}
	}
	if moduleCount == 0 {
		panic("no Go module license files were collected")
	}
	if err := os.WriteFile(filepath.Join(output, "INDEX.txt"), []byte(strings.Join(index, "\n")+"\n"), 0644); err != nil {
		panic(err)
	}
	fmt.Printf("Collected license files for %d Go modules.\n", moduleCount)
}

File: overlay/scripts/collect-licenses.py
#!/usr/bin/env python3
"""Collect license texts for files bundled into the Stratux Debian package."""

import argparse
import json
import pathlib
import re
import shutil
import subprocess


LICENSE_NAMES = ("LICENSE*", "LICENCE*", "COPYING*", "NOTICE*")


def safe_name(value: str) -> str:
    return re.sub(r"[^A-Za-z0-9._+-]+", "_", value).strip("_")


def copy_license_files(source: pathlib.Path, destination: pathlib.Path) -> list[str]:
    copied = []
    if not source.is_dir():
        return copied
    for pattern in LICENSE_NAMES:
        for path in sorted(source.glob(pattern)):
            if not path.is_file():
                continue
            target = destination / path.name
            destination.mkdir(parents=True, exist_ok=True)
            shutil.copyfile(path, target)
            copied.append(str(target))
    return copied


def decode_json_stream(value: str) -> list[dict]:
    decoder = json.JSONDecoder()
    position = 0
    records = []
    while position < len(value):
        while position < len(value) and value[position].isspace():
            position += 1
        if position >= len(value):
            break
        record, position = decoder.raw_decode(value, position)
        records.append(record)
    return records


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--source", required=True)
    parser.add_argument("--output", required=True)
    args = parser.parse_args()

    source = pathlib.Path(args.source).resolve()
    output = pathlib.Path(args.output).resolve()
    shutil.rmtree(output, ignore_errors=True)
    output.mkdir(parents=True)

    index = ["Stratux NX bundled license files", ""]

    direct_files = {
        "stratux-upstream": source / "LICENSE",
        "stratux-nx": source / "debian/stratux-nx-license",
        "stratux-nx-notices": source / "debian/STRATUX-NX-NOTICES.md",
    }
    for name, path in direct_files.items():
        if not path.is_file():
            raise SystemExit(f"Required license file is missing: {path}")
        target = output / f"{name}-{path.name}"
        shutil.copyfile(path, target)
        index.append(f"{name}: {target.name}")

    for relative in ("dump978", "dump1090", "rtl-ais", "ogn/ogn-tracker"):
        destination = output / safe_name(relative)
        copied = copy_license_files(source / relative, destination)
        if copied:
            index.append(f"{relative}: {destination.name}/")

    result = subprocess.run(
        ["go", "list", "-m", "-json", "all"],
        cwd=source,
        check=True,
        capture_output=True,
        text=True,
    )
    modules = decode_json_stream(result.stdout)
    module_count = 0
    for module in modules:
        module_path = str(module.get("Path", ""))
        module_dir = pathlib.Path(str(module.get("Dir", "")))
        if not module_path or not module_dir.is_dir() or module.get("Main"):
            continue
        destination = output / "go-modules" / safe_name(module_path)
        copied = copy_license_files(module_dir, destination)
        if copied:
            module_count += 1
            index.append(f"{module_path}: go-modules/{destination.name}/")

    if module_count == 0:
        raise SystemExit("No Go module license files were collected")

    (output / "INDEX.txt").write_text("\n".join(index) + "\n")
    print(f"Collected license files for {module_count} Go modules.")


if __name__ == "__main__":
    main()
